### PR TITLE
[CODE HEALTH] Fix clang-tidy warnings in base2 exponential histogram aggregation

### LINKED ISSUE #3979: [CODE HEALTH] clang-tidy warnings in base2_exponential_histogram_aggregation
clang-tidy reports the following warnings in `base2_exponential_histogram_aggregation.cc`
 
----

**opentelemetry-cpp/sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc** (7 warnings)

| Line | Check | Message |
|---|---|---|
| 35 | `bugprone-misplaced-widening-cast` | either cast from 'int' to 'size_t' (aka 'unsigned long') is ineffective, or there is loss of precision before the conversion |
| 215 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 327 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 328 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 329 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 412 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 413 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |

----

To reproduce locally run the following in the development container:

```
export CXX_STANDARD=14
./ci/do_ci.sh cmake.clang_tidy.test
```

The markdown report with unique errors can be generated locally after building with the `ci/create_clang_tidy_report.py` script.

### PR DESCRIPTION
**Pull request by thc1006**
Fixes #3979
Contributes to #2053

## Changes

- In `GetScaleReduction`, widen the loop condition to `int64_t` so that
  `end_index - start_index + 1` cannot overflow `int32_t` at
  `max_scale_ = 20` with extreme-span histograms, and so
  `bugprone-misplaced-widening-cast` stops firing.
- In `Downscale`, wrap the `point_data_.scale_ -= by` assignment with
  `static_cast<int32_t>(point_data_.scale_ - by)` so the
  `int32_t - uint32_t` arithmetic result is explicitly narrowed back to
  `int32_t`.
- In `Merge` and `Diff`, apply the same `static_cast<int32_t>(...)` wrap to
  the five remaining `scale_ -= scale_reduction` sites.
- Decrement the `clang-tidy.yaml` `warning_limit` by 7 for both
  `all-options-abiv1-preview` and `all-options-abiv2-preview` matrix
  entries to ratchet the limit down to the new measured count.

The arithmetic is behaviour-preserving: in the six `scale_ -= uint32_t` sites,
the existing code already relies on two's-complement wrap-around through the
unsigned intermediate, and `static_cast<int32_t>` reproduces the same bit
pattern. In `GetScaleReduction`, the `int64_t` widening gives the same
numeric result on values that fit in `int32_t` but also handles the
theoretical edge case at `max_scale_ = 20` where the bucket range can span
near `INT32_MAX`.

Existing `Base2ExponentialHistogramAggregation*` cases in
`sdk/test/metrics/aggregation_test.cc` cover all modified paths; in
particular `MergeCountInvariant` uses `max_buckets_ = 5` with five-bucket
delta cycles to deliberately exceed the limit and trigger the scale-reduction
branch in `Merge` (which calls `GetScaleReduction` and then the three
`scale_ -= scale_reduction` assignments). The `Diff` path is covered by the
scale-mismatched calls in the other Base2 tests.

Fixes the following warnings:

----

**opentelemetry-cpp/sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc** (7 warnings)

| Line | Check | Message |
|---|---|---|
| 35 | `bugprone-misplaced-widening-cast` | either cast from 'int' to 'size_t' (aka 'unsigned long') is ineffective, or there is loss of precision before the conversion |
| 215 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 327 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 328 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 329 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 412 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |
| 413 | `cppcoreguidelines-narrowing-conversions` | narrowing conversion from 'uint32_t' (aka 'unsigned int') to signed type 'int32_t' (aka 'int') is implementation-defined |

----

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc [hunk 1]
```diff
@@ -212,8 +212,8 @@ void Base2ExponentialHistogramAggregation::Downscale(uint32_t by) noexcept
     DownscaleBuckets(point_data_.negative_buckets_, by);
   }
 
-  point_data_.scale_ -= by;
-  indexer_ = Base2ExponentialHistogramIndexer(point_data_.scale_);
+  point_data_.scale_ = static_cast<int32_t>(point_data_.scale_ - by);
+  indexer_           = Base2ExponentialHistogramIndexer(point_data_.scale_);
```
[the remark this answers is not in the document]
Comment (thc1006) [reply sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc]: Good point, applied in ccdaf85a. Changed to `scale_ -= static_cast<int32_t>(by)` so the subtraction happens in the signed domain. clang-tidy still silent, existing Base2Exponential tests still green.

File: sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc [hunk 2]
```diff
@@ -324,9 +324,9 @@ std::unique_ptr<Aggregation> Base2ExponentialHistogramAggregation::Merge(
       DownscaleBuckets(high_res.positive_buckets_, scale_reduction);
       DownscaleBuckets(low_res.negative_buckets_, scale_reduction);
       DownscaleBuckets(high_res.negative_buckets_, scale_reduction);
-      low_res.scale_ -= scale_reduction;
-      high_res.scale_ -= scale_reduction;
-      result_value.scale_ -= scale_reduction;
+      low_res.scale_      = static_cast<int32_t>(low_res.scale_ - scale_reduction);
+      high_res.scale_     = static_cast<int32_t>(high_res.scale_ - scale_reduction);
+      result_value.scale_ = static_cast<int32_t>(result_value.scale_ - scale_reduction);
```
[the remark this answers is not in the document]
Comment (thc1006) [reply sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc]: Same refactor applied in ccdaf85a (`scale_ -= static_cast<int32_t>(scale_reduction)` on all three sites in Merge).

File: sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc [hunk 3]
```diff
@@ -409,8 +409,8 @@ std::unique_ptr<Aggregation> Base2ExponentialHistogramAggregation::Diff(
     {
       DownscaleBuckets(right.negative_buckets_, scale_reduction);
     }
-    left.scale_ -= scale_reduction;
-    right.scale_ -= scale_reduction;
+    left.scale_  = static_cast<int32_t>(left.scale_ - scale_reduction);
+    right.scale_ = static_cast<int32_t>(right.scale_ - scale_reduction);
```
[the remark this answers is not in the document]
Comment (thc1006) [reply sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc]: Same refactor applied in ccdaf85a on both Diff sites.

File: sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc [hunk 4]
```diff
@@ -32,7 +32,7 @@ namespace
 uint32_t GetScaleReduction(int32_t start_index, int32_t end_index, size_t max_buckets) noexcept
 {
   uint32_t scale_reduction = 0;
-  while (static_cast<size_t>(end_index - start_index + 1) > max_buckets)
+  while (static_cast<int64_t>(end_index) - start_index + 1 > static_cast<int64_t>(max_buckets))
```
[the remark this answers is not in the document]
Comment (thc1006) [reply sdk/src/metrics/aggregation/base2_exponential_histogram_aggregation.cc:35]: `max_buckets` here is `Base2ExponentialHistogramAggregationConfig::max_buckets_` (aggregation_config.h:72), a runtime-configurable `size_t` with default 160. It's a bucket count for an exponential histogram, realistic maxima are on the order of hundreds to low thousands. The theoretical overflow at values above INT64_MAX (that is, more than 9e18 buckets) isn't reachable from the public config surface. Keeping the int64_t form; it's minimum-diff and sidesteps the alternative's own corner case (`static_cast<uint64_t>(...)` of a possibly-negative int64_t span would wrap to a huge value).

File: CHANGELOG.md [hunk 5]
```diff
@@ -58,6 +58,9 @@ Increment the:
 * [CODE HEALTH] Fix misc clang-tidy warnings
   [#3993](https://github.com/open-telemetry/opentelemetry-cpp/pull/3993)
 
+* [CODE HEALTH] Fix clang-tidy warnings in base2 exponential histogram aggregation
+  [#3997](https://github.com/open-telemetry/opentelemetry-cpp/pull/3997)
```
[the remark this answers is not in the document]
Comment (thc1006) [reply CHANGELOG.md:62]: The diff hunk on this comment shows `[#3997](https://github.com/open-telemetry/opentelemetry-cpp/pull/3997)`, which is the PR URL, not an issue URL. I think the bot misread its own hunk. Nothing to change.

**Review by dbarker — APPROVED** [review]
Looks reasonable. Thanks for the warning fixes and test case!

**Review by lalitb — APPROVED** [review]
LGTM. Thanks.
